### PR TITLE
Rework prolog stack probe to save RCX and RDX in the parameter save area

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Instead of pushing and popping RCX and RDX around the stack probe, save them in their preassigned slots in the parameter save area, and then restore them.
This requires knowing the RSP-relative offset of the save area, which we compute.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
Not sure if I'll commit this just yet, just verifying the tests via the lab.

**Comment by AndyAyersMS** [comment]
@dotnet-bot retest this please

**Comment by AndyAyersMS** [comment]
Might as well try and merge this up while waiting on LLVM reviews.

@dotnet-bot retest this please

**Comment by AndyAyersMS** [comment]
Looks like we lost our test exclusions and are running all the coreclr tests. @mmitche any ideas?

**Comment by AndyAyersMS** [comment]
Test issues should be cleared up by dotnet/llilc#904

@dotnet-bot retest this please

**Comment by AndyAyersMS** [comment]
@swaroops want to review this one?

**Comment by AndyAyersMS** [comment]
@pgavlin or anyone else feel free to look too.

File: lib/Target/X86/X86FrameLowering.cpp [hunk 1]
```diff
@@ -549,54 +548,69 @@ MachineInstr *X86FrameLowering::emitStackProbeInline(
...
+  // Future optimization: don't save or restore if not live in.
   if (InProlog) {
-    BuildMI(&MBB, DL, TII.get(X86::PUSH64r)).addReg(X86::RCX);
-    BuildMI(&MBB, DL, TII.get(X86::PUSH64r)).addReg(X86::RDX);
+    // Compute the offsets.
+    X86MachineFunctionInfo *X86FI = MF.getInfo<X86MachineFunctionInfo>();
+    const int64_t CalleeSaveSize = X86FI->getCalleeSavedFrameSize();
+    const bool HasFP = hasFP(MF);
+    RCXShadowSlot = 8 + CalleeSaveSize + (HasFP ? 8 : 0);
+    RDXShadowSlot = 16 + CalleeSaveSize + (HasFP ? 8 : 0);
```
Comment (JosephTremoulet) [inline lib/Target/X86/X86FrameLowering.cpp]: Might be worth a comment for the magic 8/16 at the start here.  I presume it's the return address in the first case and additionally the rcx slot in the second case?  Might want to just define RDXShadowSlot as RCXShadowSlot + 8.

**Comment by JosephTremoulet** [comment]
LGTM

**Comment by AndyAyersMS** [comment]
Updated per feedback.

**Comment by JosephTremoulet** [comment]
update looks good, thanks

**Pull request merged**